### PR TITLE
Validate default homeserver config before loading the app

### PR DESCRIPTION
**Pull request by turt2live**
**Reviewer**: This is generally best reviewed as the final diff rather than the individual commits. Note that there's a TODO comment in the diff which has the intention of being followed up by a later PR - the PRs for this work item are getting a bit large, so this is being put up for review with a known deficiency. 

The TODO comment affects those who happen to use https://riot.im/develop/#/login?hs_url=https://matrix.org instead of the other options. It may also affect people's ability to claim 3pid validations.

See https://github.com/vector-im/riot-web/issues/9290
Fixes https://github.com/vector-im/riot-web/issues/8826

**Please review with https://github.com/matrix-org/matrix-react-sdk/pull/2941** (the implementation)

----

Implements the process described here: https://github.com/vector-im/riot-web/issues/9290#issuecomment-481966910

The expectation is that later layers (like the react-sdk) will make use of the `validated_discovery_config` option instead of interpreting the config themselves.

We intentionally block the UI from loading here to avoid races between discovery and the app loading.

~~Requires https://github.com/matrix-org/matrix-js-sdk/pull/897~~ Merged.

----

~~**Blocked on clarification regarding identity server behaviour.**~~ Going forward with this as-proposed with the intention of verifying that this is the correct behaviour.

We strive for not having defaults anymore, however the identity server is optional across the stack. Additionally, there is some concern that the federation is torn on whether or not .well-known should include an identity server:
```
Total servers: 966

Responses by discovery state
----------------------------
STATE               | COUNT | PERCENT
wk_not_found        | 463   | 47.93 %
has_identity_server | 101   | 10.46 %
no_identity_server  | 69    | 7.14 %
error               | 333   | 34.47 %
```
(sampled from #riot-web, #matrix, and #riot-web-announcements)

### CODE REVIEW

**Comment by turt2live** [comment]
asking for code review to make sure that the approach is valid. The behaviour may change though

File: README.md [hunk 1]
```diff
@@ -115,25 +115,29 @@ You can configure the app by copying `config.sample.json` to
...
-      or inviting users via email address will not work.  Matrix identity servers are
-      very simple web services which map third party identifiers (currently only email
-      addresses) to matrix IDs: see http://matrix.org/docs/spec/identity_service/unstable.html
-      for more details.  Currently the only public matrix identity servers are https://matrix.org
-      and https://vector.im.  In the future, identity servers will be decentralised.
-   * Riot will report an error if you accidentally configure both `default_server_name` _and_ `default_hs_url` since it's unclear which should take priority.
+1. `default_server_config` sets the default homeserver and identity server URL for
+   Riot to use. The object is the same as returned by [https://<server_name>/.well-known/matrix/client](https://matrix.org/docs/spec/client_server/latest.html#get-well-known-matrix-client),
+   with added support for a `server_name` under the `m.homeserver` section to display
+   a custom homeserver name. Alternatively, the config can contain a `default_server_name`
```
Comment (bwindels) [inline README.md]: Slightly confusing double meaning of "server name" here. `server_name` in `default_server_config[m.homeserver]` is for display only, and `default_server_name` is used as a hostname to get a WK config. We can't really change the name of either, but maybe make that more clear here in the README?

File: src/vector/index.js [hunk 2]
```diff
@@ -428,4 +446,63 @@ async function loadLanguage() {
...
+    }
+
+    const validatedConfig = AutoDiscoveryUtils.buildValidatedConfigFromDiscovery(serverName, result);
+
+    // Just in case we ever have to debug this
+    console.log("Using homeserver config:", validatedConfig);
+
+    // Add the newly built config to the actual config for use by the app
+    console.log("Updating SdkConfig with validated discovery information");
+    SdkConfig.add({"validated_discovery_config": validatedConfig});
```
Comment (bwindels) [inline src/vector/index.js]: discovery is maybe not very explanatory, maybe something like `validated_server_config`?

**Comment by turt2live** [comment]
Right, I'm moving forward with the identity server behaviour as proposed - if we want to change it, it isn't terribly difficult but needs some thought. This part of the whole stack is also ready for review by someone.

**Comment by turt2live** [comment]
The tests are actually fairly sensitive to these changes, so I'm pulling this out of review until I can fix them.

**Comment by almamun-dev** [comment]
Hi, one of my client needs private instant Messaging software like whatsapp. is it possible to develop riot for a private company. Please mail me, if anyone can solve my problem. Email: haydenj951@gmail.com ll skype: almamun.dev ll Whatsapp: +8801723894682

**Comment by turt2live** [comment]
Changing the base to `travis/feature/wellknown2` so that this can land with further PRs being reviewed against that branch. Will merge the feature branch to develop when it is safe.

**Comment by turt2live** [comment]
... also I'm going to press the right buttons

**Pull request merged**